fwprovider/pools/pool_membership.go

```diff
@@ -0,0 +1,300 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+package pools
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+
+	"github.com/bpg/terraform-provider-proxmox/fwprovider/attribute"
+
+	"github.com/hashicorp/terraform-plugin-framework/types"
+
+	"github.com/bpg/terraform-provider-proxmox/fwprovider/config"
+	"github.com/bpg/terraform-provider-proxmox/proxmox"
+	"github.com/bpg/terraform-provider-proxmox/proxmox/helpers/ptr"
+	"github.com/bpg/terraform-provider-proxmox/proxmox/pools"
+	proxmoxtypes "github.com/bpg/terraform-provider-proxmox/proxmox/types"
+	"github.com/hashicorp/terraform-plugin-framework-validators/resourcevalidator"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/int64planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+)
+
+var (
+	_ resource.Resource                     = (*poolMembershipResource)(nil)
+	_ resource.ResourceWithConfigure        = (*poolMembershipResource)(nil)
+	_ resource.ResourceWithImportState      = (*poolMembershipResource)(nil)
+	_ resource.ResourceWithConfigValidators = (*poolMembershipResource)(nil)
+)
+
+const (
+	mkPoolMembershipId        = "id"
+	mkPoolMembershipType      = "type"
+	mkPoolMembershipPoolId    = "pool_id"
+	mkPoolMembershipVmId      = "vm_id"
+	mkPoolMembershipStorageId = "storage_id"
+)
```

Comment:
i would prefer not extracting attribute names as constants. IMO it doesn't add much benefits, but makes the code harder to read.